open-apparel-registry/open-apparel-registry
# Add email confirmation step to user sign up

## Overview

This PR adds an email confirmation step into the account registration process. Previously user accounts were automatically activated on registering, although users still needed to sign in. Now we use allauth via django-rest-auth to generate a confirmation email with a link users must visit to confirm their accounts.

Visiting the link leads to a new ConfirmRegistration component in the React app. If the link is visited within the same session as the session in which the user has registered, then the user signs in via a session sign in. If it's visited in a different session, then the user must login. The UI will display a link to login if no session sign in succeeds but no link if the session sign in doesn't happen.

(This behvaior seems to be an opinionated setting from allauth, but we could try to override if we'd like. We discussed it a bit in Slack and it seemed odd initially but is a common-ish pattern.)

Connects #98

## Demo

<img width="958" alt="screen shot 2019-03-07 at 2 16 57 pm" src="https://user-images.githubusercontent.com/4165523/53982871-fabdb000-40e3-11e9-93b8-f282ac0691e9.png">

![Screen Shot 2019-03-08 at 4 38 25 PM](https://user-images.githubusercontent.com/4165523/54057746-ccae9d80-41c1-11e9-99d6-9b02822c2fdf.png)

![Screen Shot 2019-03-08 at 4 50 17 PM](https://user-images.githubusercontent.com/4165523/54057904-46468b80-41c2-11e9-9fcf-b1371fb21a45.png)


## Notes

In cases when a User account has been created through the Django admin or via some other mechanism that is *not* the UI registration form, the email confirmation workflow won't currently activate. I set the `.is_email_verified` method to return `True` in these cases, which means the app will treat those users as immediately confirmed.

## Testing

- get this branch, then `./scripts/update` and `./scripts/manage resetdb`
- serve the app, then register for a new account
- verify that you see the registration message depicted above and that you also see a confirmation email with a URL logged in the terminal
- open a new incognito browser and visit the account confirmation URL. When the confirmation request succeeds, verify that you see the confirmation success message but are not signed in
- visit the login page and sign in to your account to verify that that works
- register for another new account and then visit the confirmation email in the same browser session in which you registered
- verify that the confirmation succeeds and that you are signed in via session sign in
